### PR TITLE
New revpot and per-cell/segment parameters.

### LINKED ISSUE #750: Ion mechanism evaluation ordering in FVM.
For consistent treatment of user-defined ion concentration and reversal potential mechanisms, we need to:
* Update all such mechanisms strictly after any mechanisms that contribute to the corresponding ion currents.
* Check for and exclude mechanisms that write both ion currents and other ion variables.
* For the (currently in-progress) second-order FVM solver, treat ion concentration updates as split in the same way as voltages currently are; this may require also explicitly computing the ionic diffusivity in an implicit solution of the corresponding ODEs for stability.

### PR DESCRIPTION
**Pull request by halfflat**
* Collect cable cell parameter setting structures into `include/cable_cell_param.hpp`.
* Restructure electrical specifications (axial resistance, membrane capacitance) and ionic concentrations and reversal potentials on cable cells, so that these can be specified with a global default, per-cell defaults, and per-segment values.
* Allow reversal potentials to be set by a mechanism of a new kind 'revpot', which are prohibited from maintaining state or writing to any shared state other than ionic reversal potentials.
* Specify reversal potential mechanisms as global defaults or per-cell. Reversal potential mechanisms may not be specified at the level of a segment in order to avoid non-linearities arising from the discretization.
* Supply default cable cell parameter data that matches NEURON values (this is _not_ used by default).
* Replace the d_lambda calculation with one that approximates more faithfully the effect of tapered segments, and which will use the electrical values inherited by cell or global defaults.
* Supply a bundled mechanism 'nernst' that replaces the previous hard-coded Nernst potential calculation, for use (together with ion rebinding) in the new ion description scheme.
* All ions used in a cable_cell model must be present in the top level global parameter table, together with their charge.
* Extend semantics of external variables in modcc, to permit direct assignment (as opposed to just += or -=.)
* Extend `util::binary_search_index` to allow for a projection functional analagous to other range utilites.
* Add documentation on the cable cell API, sketching over many of the details, but describing in particular the interface changes for default parameters and reversal potential behaviour.

Fixes #750, #751, and #756.

### CODE REVIEW

File: arbor/include/arbor/cable_cell.hpp [hunk 1]
```diff
@@ -251,6 +219,20 @@ class cable_cell {
...
 
+    // Approximate per-segment mean attenuation b(f) at given frequency f,
+    // ignoring membrane resistance [1/µm].
+    value_type segment_mean_attenuation(value_type frequency, index_type segidx,
+        const cable_cell_parameter_set& global_defaults) const;
+
+    // Estimate of length constant λ(f) = 1/2 · 1/b(f), following
+    // Hines and Carnevale (2001), "NEURON: A Tool for Neuroscientists",
+    // Neuroscientist 7, pp. 123-135.
+    value_type segment_length_constant(value_type frequency, index_type segidx,
```
Comment (noraabiakar) [inline arbor/include/arbor/cable_cell.hpp]: Do we use this function anywhere?

File: arbor/fvm_layout.cpp [hunk 2]
```diff
@@ -59,6 +70,73 @@ namespace {
...
+
+        index_type seg_index = 0;
+        for (const auto& seg: segment_indices) {
+            const segment_info& seg_info = D.segments[seg];
+
+            if (seg_info.has_parent()) {
+                index_type cv = seg_info.parent_cv;
+
+                size_type i = parent_cv_indices.insert({cv, cv_index}).first->second;
+                if (i==cv_index) {
```
Comment (noraabiakar) [inline arbor/fvm_layout.cpp]: Are we just checking that insertion is successful here? Why not just check `parent_cv_indices.insert({cv, cv_index}).second`?

File: arbor/include/arbor/cable_cell.hpp (hunk 1 above)
Comment (halfflat) [reply arbor/include/arbor/cable_cell.hpp]: This is for use in (external) user code that uses the d_lambda measure for determining the number of CVs to use. The previous method just needed a segment, but now with the inheritable electrical properties, we need to move it out of the segment member functions and make it a cell method.

It's not currently used by any of our examples, but it probably does deserve a unit test.

File: arbor/fvm_layout.cpp (hunk 2 above)
Comment (halfflat) [reply arbor/fvm_layout.cpp]: In the instance that we've already seen this CV, we want to retrieve the previously saved index value for passing on to the action; by grabbing the value from the first field (the iterator), we can avoid a second look-up.

**Pull request merged**